cherry-pick 2.0: schema-related fixes

knz opened the pull request:
Picks #22828.
Picks #22997.
Picks #22998.
Picks #22994.
Picks #23040.
Picks #23044.

knz committed:
sql: support current_catalog as alias for current_database()

Release note (sql change): the special identifier `current_catalog` is
supported as alias for `current_database()` for better compatibility
with PostgreSQL.

knz committed:
sql: make current_role an alias for current_user

Release note (sql change): `current_role` is recognized as an alias
for `current_user` for better compatibility with PostgreSQL.

knz committed:
sql: support SET SCHEMA for better compatibility with pg and the SQL standard

At least one client app (Zippix) uses `SET SCHEMA`, defined by the SQL
standard.  `SET SCHEMA` is used by an app after discovering the
available schemas in the `information_schema` tables. Postgres
implements `SET SCHEMA` as an alias for `SET search_path =`, since the
"current schema" in pg's dialect is defined exactly as "the first item
in `search_path`".

This patch makes CockroachDB do the same as pg.

Release note (sql change): CockroachDB now recognizes the special
syntax `SET SCHEMA <name>` as an alias for `SET search_path = <name>`
for better compatibility with PostgreSQL.

knz committed:
sql: enable inspection of schemas with SHOW SCHEMAS

Now that SHOW TABLES only shows tables in the public schema by
default, and SET DATABASE only supports real databases and not
(virtual) schemas, there is no simple way for a user to realize that
there are multiple schemas in the first place, and which virtual
schemas exist / are recognized.

Release note (sql change): the new statement `SHOW SCHEMAS` reveals
which are the valid virtual schemas next to the physical schema
`public`.

knz committed:
sql: make SHOW TABLES able to inspect arbitrary schemas

With the introduction of compatibility with schema specifications,
SHOW TABLES was modified to recognize the following two forms:

     SHOW TABLES FROM <db>
     SHOW TABLES FROM <db> . <schema>

So one could, for example, issue `SHOW TABLES FROM system.pg_catalog`
to see all tables in `system`'s `pg_catalog`.

However this change broke a use case, as users were previously able to
use `SHOW TABLES` directly on virtual schemas, e.g. `SHOW TABLES FROM
crdb_internal`.

This patch restores this functionality, by making `SHOW TABLES`
resolve its parameter as if it was a star prefix in GRANT.  That is,
`SHOW TABLES FROM crdb_internal` will first try to find a database
called `crdb_internal`, and if that fails will try the current
database's schema called `crdb_internal`.

Release note (bug fix): `SHOW TABLES` is now again able to inspect
virtual schemas.

knz committed:
sql: avoid the word "pattern" in expansion error strings

Requested/suggested by @rytaft: "pattern" is a bit of a technical word
and is not necessary to convey the right meaning in the error message.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23041)
<!-- Reviewable:end -->

knz committed:
sql: clarify why it's not possible to write to virtual schemas

Prior to this patch, an attempt to populate one of the virtual schemas
with a DDL statement would fail with the error "unsupported schema
specification". This is not extremely clear, also incorrect on its
face, since the virtual schemas are actually supported.

This patch clarifies the error by reporting "schema cannot be
modified" more explicitly.

Suggested/requested by @rytaft.

Release note (sql change): attempts to modify virtual schemas with DDL
statements now fail with a clearer error message.

knz commented:
cc @cockroachdb/release

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 12 files at r4, 8 of 13 files at r5, 5 of 5 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/sql.y, line 2874 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23041#-L6CQU6QPMMOdZOnpzup:-L6CQU6QPMMOdZOnpzuq:bbdl0p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2129d010fafb5385232e5750bdbc225af3e6831/pkg/sql/parser/sql.y#L2874)):*
> ```Yacc
>     $$.val = &tree.ShowTables{TableNamePrefix:tree.TableNamePrefix{
>         // Note: the schema name may be interpreted as database name,
> 	// see name_resolution.go.
> ```

[nit] remove tab character

---

*[pkg/sql/sem/tree/show.go, line 165 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23041#-L6CQp4v77fIveqzGOwJ:-L6CQp4v77fIveqzGOwK:b-lw8zu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2129d010fafb5385232e5750bdbc225af3e6831/pkg/sql/sem/tree/show.go#L165)):*
> ```Go
> }
> 
> // ShowSchemas represents a SHOW TABLES statement.
> ```

`SHOW TABLES` -> `SHOW SCHEMAS`

---

*[pkg/sql/sem/tree/table_pattern.go, line 67 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23041#-L6CRMQcv5Xz7YMiawmW:-L6CRMQcv5Xz7YMiawmX:bfwx0ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2129d010fafb5385232e5750bdbc225af3e6831/pkg/sql/sem/tree/table_pattern.go#L67)):*
> ```Go
> func (at *AllTablesSelector) Format(ctx *FmtCtx) {
> 	at.TableNamePrefix.Format(ctx)
> 	alwaysFormatPrefix := ctx.flags.HasFlags(FmtAlwaysQualifyTableNames) || ctx.tableNameFormatter != nil
> ```

Might be good to move this line and the following `if` statement to a helper function (e.g., `func formatPrefix() bool`) since it's repeated a few times.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23041#-:-L6CRpdrswGPrpEJ0eet:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Becca I'll follow up on your requests in #23072. This PR here is merely a cherry-pick, I can't make changes here without first issuing a PR against master.
